third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1221,4 +1515,257 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+      if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+        outer_reduction_tv->split(
+            axisID, rparams.vectorization_factor_tmp_gmem_read);
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::Unroll);
+      }
+
+      if (rparams.tidx_for_outer_reduction) {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDy));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDy);
+      } else {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDx));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDx);
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    } else {
+      // reduction domain
+      outer_reduction_tv->split(
+          0, NamedScalar::getParallelDim(ParallelType::TIDy));
+      outer_reduction_tv->axis(1)->parallelize(ParallelType::TIDy);
+      // iteration domain
+      int axisID = -1;
+      if (rparams.vectorization_factor_tmp_gmem_read > 1) {
+        outer_reduction_tv->split(
+            axisID, rparams.vectorization_factor_tmp_gmem_read);
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::Unroll);
+      }
+      if (rparams.lparams.bdimx() > 1) {
+        outer_reduction_tv->split(
+            axisID, NamedScalar::getParallelDim(ParallelType::TIDx));
+        outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::TIDx);
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    }
+
+    auto outer_reference_tv =
+        reduction_scheduler_utils::sortAndRFactor(outer_reduction_tv);
+    outer_reference_tvs.emplace_back(outer_reference_tv);
+  }
+
+  for (auto output : dummy_outputs) {
+    fusion->addOutput(output);
+  }
+
+  const bool vectorize =
+      rparams.vectorize_inner_reduction || rparams.vectorize_iter_dom;
+  const bool is_outer_grid_persistence = rparams.persistent_kernel &&
+      rparams.cross_grid_inner_reduction && !rparams.fastest_dim;
+
+  // Propagate inner reduction
+  reduction_scheduler_utils::propagateTransformation(
+      inner_reference_tv, boundaryNodesSet);
+  reduction_scheduler_utils::propagateRFactor(
+      inner_reference_tv, inner_reduction_tv, inner_reduction_tvs);
+  reduction_scheduler_utils::propagateParallelization(
+      fusion,
+      inner_reduction_tv,
+      inner_reference_tv,
+      unroll,
+      vectorize,
+      is_outer_grid_persistence,
+      inner_reduction_tvs,
+      cached_inputs,
+      cached_outputs);
+
+  // Propagate outer reduction
+  for (long unsigned int i = 0; i < outer_reference_tvs.size(); i++) {
+    reduction_scheduler_utils::propagateTransformation(
```

Comment:
We have already propagated transformations from the inner reference TV. How does this propagation from the outer reference TV work? Since both are connected, does the second propagation overwrite the results by the first propagation?

The same question applies to the propagation of parallelization.

Reply:
You are right. I added a check. If they have shared consumer, will reject.
'''
      if (normalization_scheduler_utils::hasSharedConsumer(
              inner_reduction_tvs, outer_reduction_tvs)) {
        scheduler_debug_utils::canScheduleRejectReason(
            ScheduleHeuristic::Persistent,
            "to use combined reduction, inner reduction and outer reduction should NOT have shared consumer.");
        return false;
      }
'''